### DISCUSSION TITLE
Remove Alpine $watch when component disapears

### DISCUSSION THREAD

**Discussion by joffreypersia** [livewire/livewire#7546] [1. Help]
Hi, 

I am building a multi-step form with a lot of controls. 

One of them is, at each step on init() there is a $watch on a specific part of an array (that will be send at the final step)
But if the user goes to step 3 and returns to step 2, it triggers the $watch again and after, the functions related to the watch got triggered many times. 

Is there a way to isolate the $watch to the view (livewire / blade component ) displayed ? 
Is there a way to remove the $watch when the user clicks on "next" ? 

I didn't find anything related here : https://alpinejs.dev/magics/watch
and because it is caused by a new component being render by Livewire, I am posting here.

Thank you so much

**Comment by joshhanley** [comment]
@joffreypersia hard to say without seeing some of your code. Alpine should clean up watches when the alpine component is removed from the DOM, so you would need to ensure the Alpine component is contained within the step, so it gets removed.

**Comment by joffreypersia** [comment]
Thank you for your answer @joshhanley 

Here are the piece of code I am using : 
Simplified multi-step-form.blade.php

```php
<div x-data="{ isLoading: @entangle('isLoading'), isValid: @entangle('isValid') }">
    <!-- Main Steps Display (Simplified) -->
    @foreach ($steps as $stepKey => $step)
        <div>
            <span>{{ $stepIndex = array_search($stepKey, $stepKeys) + 1 }}</span>
            <span>{{ __('step-processes.'.$step['title']) }}</span>
        </div>
    @endforeach

    <!-- Dynamic Step Content -->
    <livewire:step-form :currentStepId="$currentStepId" />

    <!-- Navigation Buttons -->
    @if($currentStepId < $totalSubSteps)
        <button wire:click="decrementStep" :disabled="$currentStepId == 1">Previous</button>
        <button id="submitButton" type="submit" form="stepForm{{$currentStepId}}" :disabled="!isValid || isLoading">Next</button>
    @endif
</div>
```

Simplified MultiStepForm.php
```php
namespace App\Livewire;

use Livewire\Component;

class MultiStepForm extends Component
{
    public $currentStepId = 1;
    public $currentMainStepName;
    public $steps = [];
    public $isLoading = false;
    public $isValid = false;

    public function incrementStep(): void
    {
        // Logic to increment step
    }

    public function decrementStep(): void
    {
        // Logic to decrement step
    }

    public function render()
    {
        // Logic to prepare steps and render view
        return view('livewire.multi-step-form');
    }
}
```

And the steps views are called like step1.blade.php with mainly :
```php
<div x-data="{ 
     parameters: @entangle('parameters'), 
     source: '{{ $parameters['source'] ?? $sourceDefault }}',
     init(){
          $nextTick(() => {
                $watch('parameters', value => {this.checkIsValid()}); // This got retriggered on step2 for example
          })
    },
    checkIsValid(){
            if(this.parameters.source != null)
            {
                this.isValid = true;
            }
     },
}">
    <h4>{{ __('step-processes.export_process_step1_headline') }}</h4>
    <form id="stepForm{{$currentStepId}}" @submit.prevent="$wire.next()">
        <!-- Components -->
    </form>
</div>
```

**Reply by joshhanley** [reply]
@joffreypersia try adding a `wire:key` to your `<livewire:step-form :currentStepId="$currentStepId" />` which contains the step details. What this will do is tell Livewire to remove the old `step-form` component and create a new one, each time a step change happens. Hopefully this should clean everything up as expect. Keep in mind any state in the component will be lost if it get's removed and a new one added.

Hope this helps!
